pkg/tfgen/docs_hcl.go

```diff
@@ -0,0 +1,289 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tfgen
+
+import (
+	"github.com/hashicorp/hcl/hcl/scanner"
+	"github.com/hashicorp/hcl/hcl/token"
+)
+
+// fixHcl attempts to fix certain simple syntactical errors in a particular piece of HCL source text.
+//
+// For reference, here is the HCL grammar in ~EBNF:
+//
+//     file := objectList EOF
+//
+//     objectList := { objectItem }
+//
+//     objectItem := ( assignmentProperty | objectProperty )
+//
+//     assignmentProperty := ( IDENT | STRING ) '=' value
+//
+//     objectProperty := ( IDENT | STRING ) { ( IDENT | STRING ) } objectValue
+//
+//     value := literalValue | objectValue | listValue
+//
+//     literalValue := NUMBER | FLOAT | BOOL | STRING | HEREDOC
+//
+//     objectValue := '{' objectList '}'
+//
+//     listValue := '[' [ value { ',' value } ] ']'
+//
+// We want to fix the following errors:
+// - missing value in an assignmentProperty
+// - missing objectValue in an objectProperty
+// - missing values in a list
+//
+// We do not need to fix;
+// - unbalanced braces
+// - unbalanced brackets
+// - bad assignmentProperty keys
+// - etc.
+//
+// All of the problems we want to fix can be addressed by inserting artificial tokens into the token stream.
+func fixHcl(hcl string) (string, bool) {
+	input := []byte(hcl)
+	h := &hclFixer{
+		input:   input,
+		scanner: scanner.New(input),
+	}
+	h.scanner.Error = func(_ token.Pos, _ string) {}
+
+	if !h.file() {
+		return "", false
+	}
+
+	return string(h.output), true
+}
+
+// hclFixer is a utility type that is used by fixHcl.
+type hclFixer struct {
+	input  []byte // the input source text
+	output []byte // the output source text
+
+	scanner *scanner.Scanner // the scanner that lexes the input source text
+	next    *token.Token     // the next token to return
+
+	flushed int // the number of bytes of source text that have been processed
+}
+
+// peek returns the next token in the input text, but does not advance the current position or copy the token to the
+// output buffer.
+func (h *hclFixer) peek() token.Token {
+	// If the buffer is empty, grab the next token from the scanner.
+	if h.next == nil {
+		t := h.scanner.Scan()
+
+		// Chew through comments. We save the offset of the first comment token and stamp it onto the first non-comment
+		// token we see so that patches (if any) will occur before comments. For example, given the following:
+		//
+		//     resource "aws_s3_bucket" "foo" {
+		//         name = # put a name here
+		//     }
+		//
+		// If we patch the missing property value, we want the patch to occur before "# put a name here". If we simply
```

Comment:
Does this imply that we're dropping comments in the output stream? So your example would end up fixing to

```hcl
resource "aws_s3_bucket" "foo" {
    name = ""
}
```

Reply:
Good question! It does not. The comments are effectively treated as part of the token they precede, and will be copied to the output with that token.

The only time we may drop comments is if they precede an `ILLEGAL` token that gets eluded by `patch`.